keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1278 @@
+Job is marked Failed and its execution is terminated.
+
+We also propose to extend the PodFailurePolicy with a new action, called
+`FailIndex` to allow an index to fail fast before reaching the backoff limit
+per index.
+
+### Job API
+
+```golang
+
+// PodFailurePolicyAction specifies how a Pod failure is handled.
+// +enum
+type PodFailurePolicyAction string
+
+const (
+  // This is an action which might be taken on a pod failure - mark the
+  // Job's index as failed to avoid restarts within this index. This action
+  // can only be used when backoffLimitPerIndex=true.
+  PodFailurePolicyActionFailIndex PodFailurePolicyAction = "FailIndex"
+  ...
+)
+...
+
+// JobRunPolicy describes custom policy for running a Job
+type JobRunPolicy struct {
+
+  // Specifies the limit for the number of retries within an
+  // index before marking this index as failed. When enabled the number of
+  // failures per index is kept in the pod's
+  // batch.kubernetes.io/job-index-failure-count annotation. It can only
+  // be set to true when Job's completionMode=Indexed, and the Pod's restart
+  // policy is Never.
+  // When enabled, the maximal number of completions is limited to 10^6.
+  // The value is immutable.
+  // Defaults to false
+  // +optional
+  BackoffLimitPerIndex *int32
+
+  // Specifies the maximal number of failed indexes before marking the Job as
+  // failed, when backoffLimitPerIndex is set. Once the number of failed
+  // indexes exceeds this number the entire Job is marked as Failed and its
+  // execution is terminated. When left as nil the job continues execution of
+  // all of its indexes and is marked with the `Complete` Job condition.
+  // It can only be specified when backoffLimitPerIndex is set.
+  // The value is might be in range from 0 to 10^4.
+  // It defaults to 10^4 when completions is greater or equal than 10^4.
+  // +optional
+  MaxFailedIndexes *int32
+}
+
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
+  ...
+
+  // Specifies the policy for running a Job. In particular, it allows to enable
+  // the backoff limit per index.
+  // +optional
+  RunPolicy *JobRunPolicy
+  ...
+}
+
+type JobStatus struct {
+  ...
+
+  // FailedIndexes holds the failed indexes when backoffLimitPerIndex is set.
+  // The indexes are represented in the text format analogous as for the
+  // `completedIndexes` field, ie. they are kept as decimal integers
+  // separated by commas. The numbers are listed in increasing order. Three or
+  // more consecutive numbers are compressed and represented by the first and
+  // last element of the series, separated by a hyphen.
+  // For example, if the failed indexes are 1, 3, 4, 5 and 7, they are
+  // represented as "1,3-5,7".
+  // +optional
+  FailedIndexes *string
+}
+```
+
+Here, setting `.spec.runPolicy.backoffLimitPerIndex` along with `.spec.backoffLimit`
+results in a warning. Still, we leave it possible to set both fields to allow
```

Comment:
The use case presented by @deads2k during an offline meeting is downgrade (or just skewed versions). The user may want to set `backoffLimit` equal `backoffLimitPerIndex` * `maxFailedIndexes` * `estimated pod failure ratio` to approximate as much as possible the new behavior with the current API. Otherwise, the default would be 6, which may be not enough for many jobs.

As for the need in the initial version - I leaning towards keeping it simple and error for now. @deads2k @alculquicondor wdyt?

Reply:
Defaulting will always set `backoffLimit`, so there will always be a warning. Unless we change the defaulting to not set `backoffLimit` if `backoffLimitPerIndex` is set. But:
- defaulting depending on other fields is discouraged
- it doesn't contribute to the downgrade story: the field would be unset.

I agree that it's much better to fail validation.

/lgtm cancel
as the design changed :)

Reply:
@alculquicondor the current proposal is to default `backoffLimit` only if `backoffLimitPerIndex` isn't set. Can you point out why defaulting based on other fields is discouraged? 

> it doesn't contribute to the downgrade story: the field would be unset.

@deads2k explained to me that the defaulting happens on every read, and is persisted on the next write. Some reference for the mechanism would be good, can you provide @deads2k ?

I tested it this way: created a Job using a version when I can set `backoffLimit: nil` and `backoffLimitPerIndex: x`, then downgraded and the `backoffLimit` field was indeed set to 6. 

So, the downgrade will  be handled better than in the original proposal. In the original proposal if someone set `backoffLimit: 1` along with `backoffLimitPerIndex: true` the value of `1` will be used, which for most use-cases seem wrong.

Reply:
> Can you point out why defaulting based on other fields is discouraged?

I might be misremembering. Official docs are here:

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#defaulting

I think the concern was around what happens when the user mutates the field A that the defaulted field B depended on. The user might be doing so from a yaml that has field A but doesn't have the field B. Then if they modify field A, and re-apply the yaml, the value of field B wouldn't be modified, which could lead to a failed validation, for reasons that are not immediately obvious (the user didn't set field B, why is it failing?).
However, in you case, you are leaving the field B (backoffLimit) nil, so maybe this kind of dependent defaulting is not problematic.

> So, the downgrade will be handled better than in the original proposal. In the original proposal if someone set backoffLimit: 1 along with backoffLimitPerIndex: true the value of 1 will be used, which for most use-cases seem wrong.

Features go through an alpha release (disabled by default) for a reason. This case would only happen if the user created a Job in a cluster that had an alpha feature enabled and later downgraded. We don't need to support that scenario.

In any case, it sounds like the proposed fields and defaulting should be fine. But I would fail validation if both fields are set, instead of producing a (mostly useless) warning.

Reply:
It seems the consensus is that we should fail the validation. I will update the KEP if no objections. @deads2k wdyt?

Reply:
> I tested it this way: created a Job using a version when I can set `backoffLimit: nil` and `backoffLimitPerIndex: x`, then downgraded and the `backoffLimit` field was indeed set to 6.
> 
> So, the downgrade will be handled better than in the original proposal. In the original proposal if someone set `backoffLimit: 1` along with `backoffLimitPerIndex: true` the value of `1` will be used, which for most use-cases seem wrong.

I agree.  I think the defaulting behavior based on the PerIndex being set produces a downgrade experience more consistent with the known fields and it provides a slightly better experience for a user trying to have a single manifest to use on clusters that may or may not have this field. 


> I think the concern was around what happens when the user mutates the field A that the defaulted field B depended on. The user might be doing so from a yaml that has field A but doesn't have the field B. Then if they modify field A, and re-apply the yaml, the value of field B wouldn't be modified, which could lead to a failed validation, for reasons that are not immediately obvious (the user didn't set field B, why is it failing?).
However, in you case, you are leaving the field B (backoffLimit) nil, so maybe this kind of dependent defaulting is not problematic.

In this case, eliminating the conditional field results in exact parity with previous behavior, so the user's expectations are met.  The job_controller needs to properly honor optional fields, but optional fields are already optional so we would expect that to be true.  The same state existed for clients during the rollout of `backoffLimit` when some servers returned it and some did not.


> It seems the consensus is that we should fail the validation. I will update the KEP if no objections. @deads2k wdyt?

Doing this will make patch requests for the `backoffLimit` field behave unpredictably for older clients that aren't aware of `backoffLimitPerIndex` because both fields would be set. This amounts to tightening validation.  We can make a warning because it's not breaking, but we cannot make it hard fail and maintain compatbility with prior clients.

Reply:
> Doing this will make patch requests for the backoffLimit field behave unpredictably for older clients that aren't aware of backoffLimitPerIndex because both fields would be set.

Why would both values be set in these older clients?
Defaulting doesn't run in the client code, only in the server.

If the server returns a `nil` backoffLimit, I would expect clients to either modify backoffLimit or leave it untouched. If an old client is modifying `backoffLimit`, it probably doesn't need the semantics of `backoffLimitPerIndex`.

Reply:
I'm fine with error or warning, but leaning towards error. If we go for warning it seems it is a transient phase when old clients might be used against the new Job API. Thus, we should in that case think of promoting the warning to error for GA, or GA+k. Personally, I feel this is an unnecessary complication. @deads2k do we have a precedent for the warning strategy in a similar case?

Reply:
After an offline discussion with @alculquicondor and @deads2k we concluded to support setting both `backoffLimit` and `backoffLimitPerIndex`. First, it allows for controller downgrade, second older clients can still read and modify the backoffLimit. Also, the proposal is to default the backoffLimit to max int32 (when backoffLimitPerIndex is set), so that it is effectively not enforced.